[Cache] add CacheInterface::delete() + improve CacheTrait

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.2
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

I've been hesitating a lot on this topic, but I think we should add a `delete()` method to `CacheInterface`.
Deleting is a very common invalidation strategy and invalidating by passing `$beta=INF` to `get()` has the drawback of requiring a fetch+unserialize+save-with-past-expiration. That's complexity that a delete removes.

This PR fixes an issue found meanwhile on `get()`: passing an `ItemInterface` to its callback makes it harder than required to implement on top of PSR-6. Let's pass a `CacheItemInterface`.

Lastly, the early expiration logic can be moved from the component to the trait shipped on contracts.

Here we are for all these steps.

nicolas-grekas committed:
[Cache] add CacheInterface::delete() + improve CacheTrait

nicolas-grekas commented on src/Symfony/Contracts/Tests/Cache/CacheContractsTraitTest.php:
```diff
@@ -116,17 +121,51 @@ public function testExceptionOnNegativeBeta()
         $cache->get('key', $callback, -2);
     }
+
+    private function getPool()
```
good catch, removed

nicolas-grekas commented on src/Symfony/Contracts/Cache/CacheInterface.php:
```diff
@@ -25,17 +26,30 @@
+     *
+     * @return bool True if the item was successfully removed, false if there was any error
+     */
+    public function delete(string $key): bool;
```
Wouldn't make sense to me: having `deleteItem()` + `get()` would be inconsistent. We shouldn't care about PSR-6 CacheItemPoolInterface on this one as its totally unrelated. The (currently named) `CacheInterface` should be consistent and more importantly complete on its own. As a nice bonus, this allows providing full type hints, where PSR can't.

nicolas-grekas replied:
Oh, about the docblock, I think the explanations provide value. I'll remove if more ppl think they don't of course.

ostrolucky replied:
Regarding docblock: What I mean is that summary needs to be updated to also mention it doesn't only do get/store. And description moved to corresponding method docblock.

nicolas-grekas replied:
Got it, docblock improved.

nicolas-grekas commented on src/Symfony/Contracts/Cache/CacheContractsTrait.php:
```diff
@@ -0,0 +1,67 @@
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+trait CacheContractsTrait
```
The responsibility is mostly given by the interface this allows implementing.
The specific added responsibility here is providing an implementation of get+delete using PSR-6 methods, so that implementing the contracts interface is a oneliner.
Something like `CacheInterfaceImplementationForCacheItemPoolInterfaceTrait` would be more precise, but I fear that's not really better. At least, from a consumer pov (ie a PSR-6 implementer), having `use CacheContractsTrait;` might be a good enough hint to quickly remind one about what this is doing.
But I'd happily listen to better suggestions.

> opportunity to improve name of CacheIterface as well

Sure, naming is hard :) Any suggestions? If we really want an alternative name, my proposal would be `CacheStorageInterface`. This has the benefit of not colliding with PSR-16 `CacheInterface` (if we should care.)

ostrolucky replied:
Yep I like CacheStorage* :)

Isn't tight coupling of this trait to interface sign it would be better to replace these with abstract class?

Nyholm replied:
`CacheInterface` is simpler. Im not sure how many times I've been fighting with PSR-6 interfaces' complicated names. 
I vote for no rename of the interface. 
The trait however. Not sure... Dont care much =)

nicolas-grekas commented:
Thank you @ostrolucky and @Nyholm for the review.
I updated the docblocks and renamed a few things (I kept `CacheInterface` but the trait is now simply `CacheTrait` in contracts (and `ContractsTrait` in the component, but that's internal).

ostrolucky commented on src/Symfony/Contracts/Cache/CacheInterface.php:
```diff
@@ -11,31 +11,46 @@
- * Gets/Stores items from/to a cache.
+ * Covers most simple to advanced caching needs.
  *
  * On cache misses, a callback is called that should return the missing value.
```
This section should be moved next to get method

nicolas-grekas replied:
moved

Nyholm commented:
> I've been hesitating a lot on this topic, but I think we should add a delete() method to CacheInterface.
> Deleting is a very common invalidation strategy and invalidating by passing $beta=INF to get() has the drawback of requiring a fetch+unserialize+save-with-past-expiration. That's complexity that a delete removes.

Im not convinced this PR is needed. Doing fetch, unserialize before save is an implementation detail. 

It might be easier to use though...

nicolas-grekas commented:
> fetch, unserialize before save is an implementation detail

yes, I used that as a hint, but by the main reasoning is that it should be a common use case to invalidate by deletion. I consider I've been misled by what we do in core: we never delete. But that's because we deal append-only pools. Working with data from models, that's just usual business.
That's why I changed my mind and I think we need delete().

fabpot commented:
Thank you @nicolas-grekas.
